### ISSUE TITLE
Better Error Messages For Mapper Deserialization

### ISSUE THREAD

**Issue by jeffreyfjohnson** [Kotlin/kotlinx.serialization#454]
**What is your use-case and why do you need this feature?**
I am using `Mapper` to serialize and deserialize. Right now if I mark a field in my classes as a `Double`, but when the type of that field sent down from the SDK we use to fetch data is actually a `Long`, we get an error message along the lines of "MyClassSerializer: Long cannot be cast to Double". This is just the most simple example, it applies to any deserialization error.

**Describe the solution you'd like**
It would be very helpful to add the `currentTag` to the error message, e.g. "MyClassSerializer: Error when deserializing tag myNumber: Long cannot be cast to Double". I think I could achieve this by subclassing `Mapper` and wrapping all the `decode` calls in a `try catch` but it would be great if that was just a feature of the library.

I think this would be especially helpful because, at least if I am remembering correctly, `currentTag` will accurately reflect the nesting level of the thing it failed to decode, i.e. `myNumbers.7`. Here we would know that the value at index 7 in the array `myNumbers` was the issue.

Having this would speed up debugging significantly

**Comment by jeffreyfjohnson** [comment]
This is what I have so far. I don't have the full context on this problem, so if someone who days can guide me I am happy to expand on this and put in a PR

```
    override fun decodeTaggedUnit(tag: String): Unit = tryDecode(tag)
    override fun decodeTaggedBoolean(tag: String): Boolean = tryDecode(tag)
    override fun decodeTaggedByte(tag: String): Byte = tryDecode(tag)
    override fun decodeTaggedShort(tag: String): Short = tryDecode(tag)
    override fun decodeTaggedInt(tag: String): Int = tryDecode(tag)
    override fun decodeTaggedLong(tag: String): Long = tryDecode(tag)
    override fun decodeTaggedFloat(tag: String): Float = tryDecode(tag)
    override fun decodeTaggedDouble(tag: String): Double = tryDecode(tag)
    override fun decodeTaggedChar(tag: String): Char = tryDecode(tag)
    override fun decodeTaggedString(tag: String): String = tryDecode(tag)
    override fun decodeTaggedEnum(tag: String, enumDescription: EnumDescriptor): Int = tryDecode(tag)

    private inline fun <reified T : Any> tryDecode(tag: String): T {
        return try {
            decodeTaggedValue(tag) as T
        } catch (e: Exception) {
            throw SerializationException("Error deserializing tag $tag || ${e::class.java.simpleName}: ${e.message}", e)
        }
    }
```

I am doing this in my own `Mapper` class for now, hence the `overrides`. Sample output is 
```
SerializationException: Error deserializing tag child2List.0.grandchild2AList.0.greatGrandchild2AList.0.greatGrandchild2AString || ClassCastException: java.lang.Integer cannot be cast to java.lang.String
```

**Comment by sandwwraith** [comment]
Catching `Exception` looks a bit overkill and unrelated to the issue. You can catch `ClassCastException`s only – this class is available in the common stdlib

**Comment by qwwdfsad** [comment]
Should be partially fixed in 1.0.0-RC
